### PR TITLE
Add Thunderbird client-server multimachine test to the SLES15-SP2  #75409

### PR DESCRIPTION
**Pull request by martinsmarcelo**
Add Thunderbird client-server multimachine test to the SLES15-SP2

 Solve needles errors to the SLES15-SP1 using new  Thunderbird version
 Add Thunderbird client-server test to the SLES15SP2 - #75409
 Updates needles to run on the openSUSE Tumbleweed version
 Configured SLED NetworkManager on the setup_multimachine.pm

****Important:
    ->  certify variables in qam-mail-thunderbird test suite:
         SLE_PRODUCT=sled
    ->Schedule test "disable_screensaver" until Thunderbird tests start.

- Related ticket: https://progress.opensuse.org/issues/75409

- Needles:
  SLE =  https://gitlab.suse.de/openqa/os-autoinst-needles-sles/-/merge_requests/1467
  TW  =  https://github.com/os-autoinst/os-autoinst-needles-opensuse/pull/711

- Verification run:  
    http://10.161.228.111/group_overview/142

### CODE REVIEW

**Comment by ggkioulis** [comment]
Nice one Marcelo! 
Btw the verification runs are pending right?

**Comment by martinsmarcelo** [comment]
> Nice one Marcelo!
> Btw the verification runs are pending right?

My wrong copy/paste :( . I pasted  correct link now!

**Review by ggkioulis — APPROVED** [review]
lgtm!

**Review by okurz — CHANGES_REQUESTED** [review]
Please fix the spelling mistakes and start the git commit message subject line with capital letter.

The commit message could have a bit more details, e.g. what issue is fixed, why this needs to change, to which versions of which product it applies, link to a bug or a fate entry, the choices you made, etc. Also see https://commit.style/ or http://chris.beams.io/posts/git-commit/ as a helpful guide how to write good commit messages.

File: tests/x11/thunderbird/thunderbird_pop.pm [hunk 1]
```diff
@@ -35,6 +36,14 @@ sub run {
     }
 
     mouse_hide(1);
+    # Disable screensaver
```
Comment (dzedro) [inline tests/x11/thunderbird/thunderbird_pop.pm:39]: On non-Gnome desktop xterm will be left running.

Screensaver is disabled on multiple places in code as needed,
I think it's better to schedule `disable_screensaver` at beginning. https://openqa.suse.de/tests/5030527#step/disable_screensaver/9
turn_off_gnome_suspend should be needed only on laptops

Comment (martinsmarcelo) [reply tests/x11/thunderbird/thunderbird_pop.pm:39]: Ok, modified to user disable_screensaver. Please add on yaml scheduled. Tks

**Comment by martinsmarcelo** [comment]
> Please fix the spelling mistakes and start the git commit message subject line with capital letter.
> 
> Oh, and the "issue number" is a bit ambiguous. Better reference a complete URL

Ok, I think now I stay compliance. Thanks

File: lib/thunderbird_common.pm [hunk 2]
```diff
@@ -153,10 +158,23 @@ sub tb_send_message {
             assert_and_click "thunderbird_focus_security_exception-TW";
             assert_and_click "thunderbird_select_sentemail_window-TW";
         }
+
         assert_and_click "thunderbird_maximized_send-message";
+        if ($proto eq 'imap') {
+            #need wait few minutes to send message
+            wait_still_screen(stilltime => 580, timeout => 600, similarity_level => 45);
```
Comment (okurz) [inline lib/thunderbird_common.pm:165]: You are using `sleep()` or are simulating a sleep-like functionality. This is introducing the risk of either unnecessarily wasting time or just shifting race conditions and making them even harder to investigate. You *can* use sleep but there are only very limited cases where is the best solution (e.g. yield threads in low-level system code). Here you have the classical problem in software design of [Synchronization](https://en.wikipedia.org/wiki/Synchronization_(computer_science)) . sleep should not be used in place of proper synchronization methods. https://blogs.msmvps.com/peterritchie/2007/04/26/thread-sleep-is-a-sign-of-a-poorly-designed-program/ is one of multiple references that explains this: "The original problem is likely a timing/synchronization issue, ignoring it by hiding it with Thread.Sleep is only going to delay the problem and make it occur in random, hard to reproduce ways." And if the problem does not occur at least time is wasted.

We are waiting because some condition changes some time ... keyword(s) is/are *some time*. It is important to know the exact condition that the code should wait for and check for that instead of delaying program execution for an arbitrary amount of time. For openQA tests in general just try to do what you as a human user would also do. You would not look on your watch waiting for exactly 100s before you look at the screen again, right? :)

File: lib/thunderbird_common.pm [hunk 3]
```diff
@@ -153,10 +158,23 @@ sub tb_send_message {
...
+            assert_and_click "thunderbird_dont_save";
+            assert_and_click "thunderbird_select_email_reply";
+            assert_and_click "thunderbird_reply_email";
+            assert_and_click "thunderbird_maximized_send-message";
+        }
+        assert_screen('thunderbird_sent-folder-appeared', timeout => 180);
     }
     else {
-        assert_screen 'thunderbird_sent-folder-appeared';
+        assert_screen('thunderbird_sent-folder-appeared', 90);
```
Comment (okurz) [inline lib/thunderbird_common.pm:177]: no need to add parentheses. Just write `assert_screen 'thunderbird_sent-folder-appeared', 90;`

File: lib/thunderbird_common.pm [hunk 4]
```diff
@@ -173,7 +191,7 @@ sub tb_check_email {
     wait_screen_change { send_key "shift-f5" };
     send_key "ctrl-shift-k";
     wait_screen_change { type_string "$mail_search" };
-    assert_screen "thunderbird_sent-message-received";
+    assert_screen("thunderbird_sent-message-received", timeout => 120);
```
Comment (okurz) [inline lib/thunderbird_common.pm:194]: better be consistent, just write `assert_screen 'thunderbird_sent-message-received', 120;`

File: tests/x11/thunderbird/thunderbird_imap.pm [hunk 5]
```diff
@@ -33,21 +34,29 @@ sub run {
 
     if ($hostname eq 'client') {
         $account = "internal_account_C";
+        # for imap  version, need disable screensaver/suspend
```
Comment (okurz) [inline tests/x11/thunderbird/thunderbird_imap.pm:37]: you say imap version but then specify gnome. Meant to say "gnome"? Also a duplicate space after "imap"

**Review by okurz — CHANGES_REQUESTED** [review]

File: lib/thunderbird_common.pm [hunk 6]
```diff
@@ -82,8 +82,13 @@ sub tb_setup_account {
         assert_and_click "thunderbird_SSL_advanced_config";
         assert_and_click "thunderbird_SSL_ok_config";
         assert_and_click "thunderbird_skip-system-integration";
-        assert_and_click "thunderbird_confirm_security_exception";
+        assert_and_click "thunderbird_close_advanced_settings";
+        if (is_sle('15-SP2+')) {
+            assert_and_click "thunderbird_SSL_close_privace_notes";
```
[the remark this answers is not in the document]
Comment (martinsmarcelo) [reply lib/thunderbird_common.pm:87]: ok, solved. tks

File: lib/thunderbird_common.pm (hunk 2 above)
Comment (martinsmarcelo) [reply lib/thunderbird_common.pm:165]: I understand the point, thanks, Working on that

File: lib/thunderbird_common.pm [hunk 7]
```diff
@@ -82,8 +82,13 @@ sub tb_setup_account {
...
+        }
```
[the remark this answers is not in the document]
Comment (martinsmarcelo) [reply lib/thunderbird_common.pm:88]: ok, done!

File: lib/thunderbird_common.pm (hunk 2 above)
Comment (martinsmarcelo) [reply lib/thunderbird_common.pm:165]: ok, done! Tks.

File: lib/thunderbird_common.pm (hunk 3 above)
Comment (martinsmarcelo) [reply lib/thunderbird_common.pm:177]: ok, done

File: lib/thunderbird_common.pm (hunk 4 above)
Comment (martinsmarcelo) [reply lib/thunderbird_common.pm:194]: ok, done

File: tests/x11/thunderbird/thunderbird_imap.pm (hunk 5 above)
Comment (martinsmarcelo) [reply tests/x11/thunderbird/thunderbird_imap.pm:37]: ok, done

File: lib/thunderbird_common.pm [hunk 8]
```diff
@@ -136,14 +135,15 @@ sub tb_send_message {
...
-            assert_and_click "thunderbird_confirm_security_exception";
             # Now, return focus to thunderbirt sent email window.
             hold_key "alt";
             send_key "f1";
             release_key "alt";
             send_key "tab";
             send_key "tab";
             send_key "ret";
+            assert_and_click "thunderbird_confirm_security_exception";
+            #assert_and_click "pare_aqui";
```
Comment (okurz) [inline lib/thunderbird_common.pm:146]: Don't add disabled code as nobody but you will understand why this is not active. Better leave it out and keep in your local git repository, either in `git stash` or a temporary "WIP"-commit

File: lib/thunderbird_common.pm [hunk 9]
```diff
@@ -153,10 +153,12 @@ sub tb_send_message {
             assert_and_click "thunderbird_focus_security_exception-TW";
             assert_and_click "thunderbird_select_sentemail_window-TW";
         }
+
         assert_and_click "thunderbird_maximized_send-message";
+        assert_screen "thunderbird_sent-folder-appeared", 180;
     }
     else {
-        assert_screen 'thunderbird_sent-folder-appeared';
+        assert_screen "thunderbird_sent-folder-appeared", 90;
```
Comment (okurz) [inline lib/thunderbird_common.pm:161]: please keep single ticks by default unless you want to expand perl variables within strings

File: tests/x11/thunderbird/thunderbird_imap.pm [hunk 10]
```diff
@@ -39,15 +40,16 @@ sub run {
     }
 
     mouse_hide(1);
+
     # clean up and start thunderbird
     x11_start_program("xterm -e \"killall -9 thunderbird; find ~ -name *thunderbird | xargs rm -rf;\"", valid => 0);
-    my $success = eval { x11_start_program("thunderbird", match_timeout => 300); 1 };
+    my $success = eval { x11_start_program("thunderbird", match_timeout => 120); 1 };
```
Comment (okurz) [inline tests/x11/thunderbird/thunderbird_imap.pm:46]: I don't see the reason for this change. I assume the rather high timeout was selected in the past based on observations where tests failed due to this. Why lower it?

File: tests/x11/thunderbird/thunderbird_pop.pm [hunk 11]
```diff
@@ -35,6 +36,7 @@ sub run {
     }
 
     mouse_hide(1);
+
```
Comment (okurz) [inline tests/x11/thunderbird/thunderbird_pop.pm:39]: unrelated change, please revert

File: tests/x11/thunderbird/thunderbird_pop.pm [hunk 12]
```diff
@@ -22,6 +22,7 @@ use strict;
 use testapi;
 use utils;
 use base "thunderbird_common";
+use x11utils qw(ensure_unlocked_desktop turn_off_gnome_screensaver turn_off_gnome_suspend);
```
Comment (okurz) [inline tests/x11/thunderbird/thunderbird_pop.pm:25]: you are not introducing any new uses of these functions, why import?

**Review by okurz — CHANGES_REQUESTED** [review]
See inline comments

Also in your git commit message the line breaks seem to have gone wrong. 

Please also phrase your git commit message in imperative mood.

Also see https://commit.style/ or http://chris.beams.io/posts/git-commit/ as a helpful guide how to write good commit messages.

Please also keep in mind that the github pull request description is only visible on github, the commit can be considered permanent information storage.

I can recommend the tool [hub](https://github.com/github/hub) (zypper in rubygem-hub) for easier PR creation. Also, I myself use a script [pr-last](http://github.com/okurz/scripts/tree/master/pr-last) to create a PR with proper description for these simple one-commit PRs:
```bash
$ cat $(which git-pr-last )
#!/bin/sh -e
target="${target:-"$USER"}"
git push $target && git show --no-patch --format=%B | hub pull-request -F -
```

File: lib/thunderbird_common.pm (hunk 9 above)
Comment (martinsmarcelo) [reply lib/thunderbird_common.pm:161]: ok, I will keep in mind to use simple ticks. tks

File: lib/thunderbird_common.pm (hunk 8 above)
Comment (martinsmarcelo) [reply lib/thunderbird_common.pm:146]: OMG, that comment completely escaped :(

File: tests/x11/thunderbird/thunderbird_imap.pm (hunk 10 above)
Comment (martinsmarcelo) [reply tests/x11/thunderbird/thunderbird_imap.pm:46]: The high value was used on my lab,  because my openqa lab have poor resources to run. Correct is 120.

File: tests/x11/thunderbird/thunderbird_pop.pm (hunk 11 above)
Comment (martinsmarcelo) [reply tests/x11/thunderbird/thunderbird_pop.pm:39]: OK.

File: tests/x11/thunderbird/thunderbird_pop.pm (hunk 12 above)
Comment (martinsmarcelo) [reply tests/x11/thunderbird/thunderbird_pop.pm:25]: Was used in other moment, removed. My lasts commits was not very happy, really.

**Review by okurz — COMMENTED** [review]
I still don't understand the git commit message. You say "Add Thunderbird to S15SP2" but I do not see 15SP2 mentioned anywhere

**Review by okurz — COMMENTED** [review]
What about a corresponding TW test? Because of [Factory First](https://opensource.suse.com/suse-open-source-policy) we should test against Tumbleweed by default (with exceptions for special packages, of course). Could you please check if it works with a current TW snapshot and then add to the main.pm there?

See https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/CONTRIBUTING.md#coding-style for more details

**Comment by martinsmarcelo** [comment]
> What about a corresponding TW test? Because of [Factory First](https://opensource.suse.com/suse-open-source-policy) we should test against Tumbleweed by default (with exceptions for special packages, of course). Could you please check if it works with a current TW snapshot and then add to the main.pm there?
> 
> See https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/CONTRIBUTING.md#coding-style for more details

Please verify TW tests on http://10.161.228.111/

**Comment by martinsmarcelo** [comment]
> I still don't understand the git commit message. You say "Add Thunderbird to S15SP2" but I do not see 15SP2 mentioned anywhere

ok, I think now I more clear.
Thanks

File: tests/network/setup_multimachine.pm [hunk 13]
```diff
@@ -41,6 +41,12 @@ sub run {
             disable_and_stop_service('NetworkManager', ignore_failure => 1);
             assert_script_run 'systemctl start  wicked';
         }
+        #if server runing on SLED
```
Comment (dzedro) [inline tests/network/setup_multimachine.pm:44]: Why are you switching from `NetworkManager` to `wicked` ?
`NetworkManager` is default on SLED and any test should run with it.

File: lib/thunderbird_common.pm [hunk 14]
```diff
@@ -73,16 +73,17 @@ sub tb_setup_account {
                 send_key 'down';
                 send_key 'ret';
             }
-            assert_screen "thunderbird_wizard-$proto-selected";
+            if (is_sle) {
```
Comment (okurz) [inline lib/thunderbird_common.pm:76]: How can this be a SLE specific behavior? isn't it rather depending on version only?

Comment (martinsmarcelo) [reply lib/thunderbird_common.pm]: Ok, I think now is better. Thanks

File: tests/network/setup_multimachine.pm (hunk 13 above)
Comment (martinsmarcelo) [reply tests/network/setup_multimachine.pm:43]: ok, I change to configure SLED NetwokrManager . Thanks

File: tests/network/setup_multimachine.pm [hunk 15]
```diff
@@ -36,10 +36,16 @@ sub run {
...
-        #if server running opensuse.
+        #if server running openSUSE.
         if (is_opensuse) {
-            disable_and_stop_service('NetworkManager', ignore_failure => 1);
-            assert_script_run 'systemctl start  wicked';
+            assert_script_run 'systemctl restart  wicked';
+        }
+        #if server runing on SLED
+        my $base_product = get_var('SLE_PRODUCT');
+        if ($base_product eq "sled") {
```
Comment (okurz) [inline tests/network/setup_multimachine.pm:45]: This would likely lead to perl warnings about comparison against undefined variables. Better use

```suggestion
        if (check_var('SLE_PRODUCT', 'sled')) {
```

but in general I doubt it's a save approach to rely on a specific product brand to derive if wicked or NetworkManager is used.

Comment (martinsmarcelo) [reply tests/network/setup_multimachine.pm]: Ok, sounds good. 

I think maybe a new ticket to verify this, a better approach to this situation with product brand and network.

**Review by dzedro — APPROVED** [review]

File: tests/network/setup_multimachine.pm [hunk 16]
```diff
@@ -36,30 +36,33 @@ sub run {
...
     if ($hostname =~ /server|master/) {
         setup_static_mm_network('10.0.2.101/24');
-        #if server running opensuse.
+        #if server running openSUSE.
         if (is_opensuse) {
-            disable_and_stop_service('NetworkManager', ignore_failure => 1);
-            assert_script_run 'systemctl start  wicked';
+            assert_script_run 'systemctl restart  wicked';
+        }
+        #if server runing on SLED
```
Comment (okurz) [inline tests/network/setup_multimachine.pm:43]: typo: s/runing/running/. And I suggest to follow the common style to have a space for separation after `#` and start uppercase. But actually in this case you might consider leaving out the comment, IMHO the code itself is readable enough :)

Comment (martinsmarcelo) [reply tests/network/setup_multimachine.pm]: Aff! Ok, changed. And double chek about # and Uppercas to commom style. Thanks

**Review by okurz — APPROVED** [review]
